### ISSUE TITLE
Need a help

### ISSUE THREAD

**Issue by selwynshen** [alturkovic/distributed-lock#20]
Dear Alen,

I've made a distributed-lock spring boot starter, replaced Lock implementation to redisson, but when i integrated it to my example project, it failed with an error which confused me a lot. 
It seems that if LockBeanPostProcessor is instantiated before WebMvcAutoConfiguration, then this error would occur. And also, I've tried excluding AutoConfiguration from spring.factories files and using @EnableXXX, then it will be OK, i.e. LockBeanPostProcessor instance should be instantiated after WebMvcAutoConfiguration loaded.
From you doc, @EnableJdbcDistributedLock would be annotated in project, i wonder if this is the same situation.
I wanna if you could help me find a solution. Thx.

##### The components used
- spring-boot-starter-web 2.3.7
- redisson-spring-boot-starter 3.14.1
- distributed-lock 1.4.0

##### The error before server started
```java
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'resourceHandlerMapping' defined in class path resource [org/springframework/boot/autoconfigure/web/servlet/WebMvcAutoConfiguration$EnableWebMvcConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.web.servlet.HandlerMapping]: Factory method 'resourceHandlerMapping' threw exception; nested exception is java.lang.IllegalStateException: No ServletContext set
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:657) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:637) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1336) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1176) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:556) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:516) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:324) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:322) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:897) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:879) ~[spring-context-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:551) ~[spring-context-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:143) ~[spring-boot-2.3.7.RELEASE.jar:2.3.7.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:758) [spring-boot-2.3.7.RELEASE.jar:2.3.7.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:750) [spring-boot-2.3.7.RELEASE.jar:2.3.7.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:405) [spring-boot-2.3.7.RELEASE.jar:2.3.7.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315) [spring-boot-2.3.7.RELEASE.jar:2.3.7.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1237) [spring-boot-2.3.7.RELEASE.jar:2.3.7.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) [spring-boot-2.3.7.RELEASE.jar:2.3.7.RELEASE]
	at cn.hrfax.redisson.RedissonApplication.main(RedissonApplication.java:17) [classes/:na]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.web.servlet.HandlerMapping]: Factory method 'resourceHandlerMapping' threw exception; nested exception is java.lang.IllegalStateException: No ServletContext set
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:652) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	... 20 common frames omitted
Caused by: java.lang.IllegalStateException: No ServletContext set
	at org.springframework.util.Assert.state(Assert.java:76) ~[spring-core-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at org.springframework.web.servlet.config.annotation.WebMvcConfigurationSupport.resourceHandlerMapping(WebMvcConfigurationSupport.java:534) ~[spring-webmvc-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_261]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_261]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_261]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_261]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.2.12.RELEASE.jar:5.2.12.RELEASE]
	... 21 common frames omitted
```

##### Part of codes
![image](https://user-images.githubusercontent.com/3644714/117252256-4614aa00-ae78-11eb-896b-693515fd8bfd.png)
![image](https://user-images.githubusercontent.com/3644714/117252376-72c8c180-ae78-11eb-8a1d-ee7fd11a0986.png)

Sincerely
Selwyn

**Comment by alturkovic** [comment]
I am not sure I can help with the current context you provided. The `distributed-lock-example` uses spring-web and distributed-lock libraries without any issues I am aware of.

How does your `Lock` implementation with Redisson look like? 
Did you import `DistributedLockConfiguration` using `EnableDistributedLock` or with your own alias?

It would be best if you could create a minimal project replicating the issue and posting a link to it here.

**Comment by selwynshen** [comment]
OK, I'll create a demo project and post to you a little while.

**Comment by selwynshen** [comment]
Sorry that it takes me some time here in China.
[dl-demo](https://github.com/selwynshen/dl-demo)

**Comment by selwynshen** [comment]
By the way, `distributed-lock-example` is OK because it uses `@EnableDistributedLock` which can ensure spring mvc started first and then distributed-lock configuration be loaded.

**Comment by alturkovic** [comment]
Sorry for the late response.

There seems to be an issue with the ordering of bean creation for `ConversionService`.

Adding a custom bean in the `dl-example` projects helps resolve this using:
```
@Configuration
public class ConversionServiceConfiguration {

    @Bean
    public ConversionService conversionService() {
        return new GenericConversionService();
    }
}
```

I might need to add `@AutoConfigureAfter(WebMvcAutoConfiguration.class)` but I need to test that, I am currently a bit short on time unfortunately so I don't know when I will get the chance.

**Comment by selwynshen** [comment]
Thx a lot. I can achieve it myself.

**Comment by selwynshen** [comment]
Seems that I've tried @AutoConfigureAfter(WebMvcAutoConfiguration.class) several days before, but still failed. I'll confirm it next weekdays.

**Comment by alturkovic** [comment]
Sorry, I meant I should include it in `distributed-lock-core` so you don't have to :)

**Comment by selwynshen** [comment]
I found that if `@ConditionalOnMissingBean` annotation wasn't defined on `ConversionService` bean, then it works. Seems that if webmvc is integrated, then it fails because of the injection of `mvcConversionService`.

I've created a new branch `webmvc` for my demo project `dl-demo`. It worked OK since i defined a  specific `conversionService` bean and injected to `SpelKeyGenerator`.

Thanks a lot for your help.

**Comment by alturkovic** [comment]
But what about others who do not use webmvc? The library should provide a default `ConversionService` itself. The problem is in the `DistributedLockConfiguration` class, I'll see if I get some time for the weekend to fix and test it.

**Comment by selwynshen** [comment]
You're right. But the problem is if other `ConversionService` bean defined, then distributed-lock do not work well.

When we use both spring-boot-web and concul, there even will be 2 `ConversionService` bean.

I've rewrited `DistributedLockConfiguration` class. Here is my code snippet. Hope for your suggestions.
```java
@Bean("conversionService")
@ConditionalOnMissingBean(name = "conversionService")
public ConversionService conversionService() {
    return new DefaultConversionService();
}

@Bean("keyGenerator")
@ConditionalOnMissingBean(name = "keyGenerator")
@ConditionalOnBean(ConversionService.class)
public KeyGenerator keyGenerator(@Qualifier("conversionService") final ConversionService conversionService) {
    return new StringKeyGenerator(conversionService);
}
```

**Comment by alturkovic** [comment]
Sorry for the late response again. I released a new version (1.4.2) which should resolve your issue. Please, let me know.

**Comment by selwynshen** [comment]
Thanks very much. Since I used `Spring Boot 2.3.7.Release`, I won't integrate 1.4.2 to my project, but your resolution is pretty good. I've just rewrited a new configuration class to replace the old one and it works well.
